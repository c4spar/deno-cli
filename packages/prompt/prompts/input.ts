import { blue } from 'https://deno.land/std@0.63.0/fmt/colors.ts';
import { Figures } from '../lib/figures.ts';
import { GenericInput, GenericInputKeys, GenericInputPromptOptions, GenericInputPromptSettings } from '../lib/generic-input.ts';

export interface InputKeys extends GenericInputKeys {}

export interface InputOptions extends GenericInputPromptOptions<string> {
    minLength?: number;
    maxLength?: number;
    keys?: InputKeys;
}

interface InputSettings extends GenericInputPromptSettings<string> {
    minLength: number;
    maxLength: number;
    keys?: InputKeys;
}

export class Input extends GenericInput<string, InputSettings> {

    public static async prompt( options: string | InputOptions ): Promise<string> {

        if ( typeof options === 'string' ) {
            options = { message: options };
        }

        return new this( {
            pointer: blue( Figures.POINTER_SMALL ),
            minLength: 0,
            maxLength: Infinity,
            ...options
        } ).prompt();
    }

    protected validate( value: string ): boolean | string {
        if ( typeof value !== 'string' ) {
            return false;
        }
        if ( value.length < this.settings.minLength ) {
            return `Value must be longer then ${ this.settings.minLength } but has a length of ${ value.length }.`;
        }
        if ( value.length > this.settings.maxLength ) {
            return `Value can't be longer then ${ this.settings.maxLength } but has a length of ${ value.length }.`;
        }
        return true;
    }

    protected transform( value: string ): string | undefined {
        return value.trim();
    }

    protected format( value: string ): string {
        return value;
    }
}
