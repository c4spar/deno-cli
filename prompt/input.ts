import { blue } from "./deps.ts";
import { Figures } from "./figures.ts";
import {
  GenericInput,
  GenericInputKeys,
  GenericInputPromptOptions,
  GenericInputPromptSettings,
} from "./_generic_input.ts";

export type InputKeys = GenericInputKeys;

/** Input prompt options. */
export interface InputOptions extends GenericInputPromptOptions<string> {
  minLength?: number;
  maxLength?: number;
  keys?: InputKeys;
}

/** Input prompt settings. */
interface InputSettings extends GenericInputPromptSettings<string> {
  minLength: number;
  maxLength: number;
  keys?: InputKeys;
}

/** Input prompt representation. */
export class Input extends GenericInput<string, InputSettings> {
  /** Execute the prompt and show cursor on end. */
  public static async prompt(options: string | InputOptions): Promise<string> {
    if (typeof options === "string") {
      options = { message: options };
    }

    return new this({
      pointer: blue(Figures.POINTER_SMALL),
      minLength: 0,
      maxLength: Infinity,
      ...options,
    }).prompt();
  }

  /**
   * Validate input value.
   * @param value User input value.
   * @return True on success, false or error message on error.
   */
  protected validate(value: string): boolean | string {
    if (typeof value !== "string") {
      return false;
    }
    if (value.length < this.settings.minLength) {
      return `Value must be longer then ${this.settings.minLength} but has a length of ${value.length}.`;
    }
    if (value.length > this.settings.maxLength) {
      return `Value can't be longer then ${this.settings.maxLength} but has a length of ${value.length}.`;
    }
    return true;
  }

  /**
   * Map input value to output value.
   * @param value Input value.
   * @return Output value.
   */
  protected transform(value: string): string | undefined {
    return value.trim();
  }

  /**
   * Format output value.
   * @param value Output value.
   */
  protected format(value: string): string {
    return value;
  }
}
