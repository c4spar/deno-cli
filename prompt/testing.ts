import { AssertionError } from "https://deno.land/std@0.170.0/testing/asserts.ts";
import { assertSnapshot } from "https://deno.land/std@0.170.0/testing/snapshot.ts";
import { eraseDown } from "../ansi/ansi_escapes.ts";
import { basename } from "./deps.ts";

export interface AssertPromptSnapshotOptions {
  /** Test name. */
  name: string;
  /** Import meta. Required to determine the import url of the test file. */
  meta: ImportMeta;
  /**
   * Object of test steps. Key is the test name and the value is an array of
   * input sequences/characters.
   */
  steps: Record<string, Array<string>>;
  /** Test function. */
  fn(): Promise<unknown>;
  /**
   * Operating system snapshot suffix. This is useful when your test produces
   * different output on different operating systems.
   */
  osSuffix?: Array<typeof Deno.build.os>;
  /**
   * Arguments passed to the `deno test` command when executing the snapshot
   * tests. `--allow-env=__CLIFFY_TEST_NAME__` is passed by default.
   */
  args?: Array<string>;
  /**
   * Enable/disable colors. Default is `false`.
   */
  colors?: boolean;
}

const encoder = new TextEncoder();

/**
 * Run prompt snapshot tests.
 *
 * ```ts
 * import { assertPromptSnapshot } from "./testing.ts";
 * import { Input } from "./input.ts";
 *
 * await assertPromptSnapshot({
 *   name: "test name",
 *   meta: import.meta,
 *   osSuffix: ["windows"],
 *   args: [],
 *   steps: {
 *     "should enter som text": ["foo bar", "\n"],
 *   },
 *   async fn() {
 *     await Input.prompt({
 *       message: "Whats your name?",
 *       default: "foo",
 *     });
 *   },
 * });
 * ```
 *
 * @param options Test options
 */
export async function assertPromptSnapshot(
  options: AssertPromptSnapshotOptions,
): Promise<void> {
  if (options.meta.main) {
    const testName = Deno.env.get("__CLIFFY_TEST_NAME__");
    if (testName === options.name) {
      await options.fn();
    }
  } else {
    registerTest(options);
  }
}

function registerTest(options: AssertPromptSnapshotOptions) {
  const fileName = basename(options.meta.url);

  Deno.test({
    name: options.name,
    async fn(ctx) {
      const tests = Object.entries(options.steps ?? {});
      if (!tests.length) {
        throw new Error(`No tests defined for: ${options.meta.url}`);
      }

      for (const [name, inputs] of tests) {
        await ctx.step({
          name,
          async fn(stepCtx) {
            const output: string = await runPrompt(inputs, options);
            const suffix = options.osSuffix?.includes(Deno.build.os)
              ? `.${Deno.build.os}`
              : "";

            await assertSnapshot(stepCtx, output, {
              path: `__snapshots__/${fileName}${suffix}.snap`,
            });
          },
        });
      }
    },
  });
}

async function runPrompt(
  inputs: Array<string>,
  options: AssertPromptSnapshotOptions,
): Promise<string> {
  try {
    const cmd = new Deno.Command("deno", {
      stdin: "piped",
      stdout: "piped",
      stderr: "piped",
      args: [
        "run",
        "--allow-env=__CLIFFY_TEST_NAME__",
        ...options.args ?? [],
        options.meta.url,
      ],
      env: {
        __CLIFFY_TEST_NAME__: options.name ?? "",
        NO_COLOR: options.colors ? "false" : "true",
      },
    });
    const child: Deno.ChildProcess = cmd.spawn();
    const writer = child.stdin.getWriter();

    for (const input of inputs) {
      await writer.write(encoder.encode(input));
      // Ensure all inputs are processed and rendered separately.
      await new Promise((resolve) => setTimeout(resolve, 600));
    }

    const { success, stdout, stderr, code } = await child.output();
    writer.releaseLock();
    await child.stdin.close();

    if (!success) {
      throw new Error(
        `Test command failed with a none zero exit code: ${code}. ${stderr}`,
      );
    }

    // Add a line break after each test input.
    return "stdout:\n" + new TextDecoder().decode(stdout).replaceAll(
      eraseDown(),
      eraseDown() + "\n",
    ) + "\nstderr:\n" + new TextDecoder().decode(stderr).replaceAll(
      eraseDown(),
      eraseDown() + "\n",
    );
  } catch (error: unknown) {
    const assertionError = new AssertionError(
      `Test failed: ${options.meta.url}`,
    );
    assertionError.cause = error;
    throw error;
  }
}
