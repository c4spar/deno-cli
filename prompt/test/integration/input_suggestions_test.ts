import { Input } from "../../input.ts";
import { assertPromptSnapshot } from "../../testing.ts";

await assertPromptSnapshot({
  name: "input prompt with suggestions",
  meta: import.meta,
  osSuffix: ["windows"],
  steps: {
    "should enable suggestions and list": ["foo", "\n"],
  },
  async fn() {
    await Input.prompt({
      message: "Whats your name?",
      default: "foo",
      suggestions: ["foo", "bar", "baz"],
      list: true,
    });
  },
});
