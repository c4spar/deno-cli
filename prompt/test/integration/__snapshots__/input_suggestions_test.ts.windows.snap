export const snapshot = {};

snapshot[`input prompt with suggestions > should enable suggestions and list 1`] = `
"stdout:
? Whats your name? (foo) » 
  foo
  bar
  baz\\x1b[3A\\x1b[28G\\x1b[G\\x1b[0J
? Whats your name? (foo) » foo
❯ foo\\x1b[1A\\x1b[31G\\x1b[G\\x1b[0J
? Whats your name? (foo) » foo
\\x1b[?25h\\x1b[?25h
stderr:
"
`;
